pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -150,6 +151,49 @@ func normalizeClusters(push *model.PushContext, proxy *model.Proxy, clusters []*
 	return out
 }
 
+// autoFillMTLSSettings adds service-level (m)TLS settings to the given destination rule, if applicable
+// (destination service can accept mTLS). The function returns the origin destination rule if it is already
+// has (service-level) TLS settings.
+// This function should be called for each service port, with the mTLS mode for that service + port.
+func autoFillMTLSSettings(config *model.Config, externalService bool,
+	destServiceMTLSMode authn_v1alpha1_applier.MutualTLSMode) *networking.DestinationRule {
+	destinationRule := &networking.DestinationRule{}
+	if config != nil {
+		destinationRule = config.Spec.(*networking.DestinationRule)
+		if destinationRule.TrafficPolicy != nil && destinationRule.TrafficPolicy.Tls != nil {
+			// Destination rule already has TLS settings, do nothing and return the DR as is.
+			// Note: technically, if DR has TLS settings in port-level or subset, we should returns the original as well.
+			// However, as port-level (or subset) will override the service-level, the code is simpler to add them regardless.
+			return destinationRule
+		}
+	}
+
+	if externalService {
+		// If service is (mesh)External, return DR as is.
+		return destinationRule
+	}
+
+	// The destination service is eligible to use mTLS if
+	// - The (authentication) mTLS mode is STRICT.
+	// - The (authentication) mTLS mode is PERMISSIVE, all service instances has sidecar and the autoMtls is enabled.
+	// TODO(diemvu): handle MTLSPermissive case.
+	eligibleForMTLS := (destServiceMTLSMode == authn_v1alpha1_applier.MTLSStrict)
+	if !eligibleForMTLS {
+		// Not eligible for mTLS, returns destination rule as is.
+		return destinationRule
+	}
+
+	// Injecting service-level TLS settings for ISTIO_MUTUAL. Initialize its traffic policy if needed.
+	if destinationRule.TrafficPolicy == nil {
+		destinationRule.TrafficPolicy = &networking.TrafficPolicy{}
+	}
+	destinationRule.TrafficPolicy.Tls = &networking.TLSSettings{
+		Mode: networking.TLSSettings_ISTIO_MUTUAL,
+	}
```

Comment:
this will end up modifying an existing dest rule which is not good because dest rules are shared

Reply:
Yeah, I've realized this while adding few more tests. The previous implementation was actually better in this part. I guess I need to change to carry the `smart-default TLS settings` in the build ops.

Reply:
Fixed. PTAL